Mathlib/CategoryTheory/Category/FreeGroupoid.lean

```diff
@@ -0,0 +1,211 @@
+`C ⥤ Quotient (FreeGroupoid.homRel C)`. -/
+protected def FreeGroupoid := Quotient (FreeGroupoid.homRel C)
+
+variable {C} in
+instance [Nonempty C] : Nonempty (Category.FreeGroupoid C) := by
+  have : Inhabited (CategoryTheory.FreeGroupoid C) := by
+    inhabit CategoryTheory.FreeGroupoid C
+    exact ⟨@default (CategoryTheory.FreeGroupoid C) _⟩
+  have : Inhabited (Category.FreeGroupoid C) := inferInstanceAs (Inhabited <| Quotient _)
+  inhabit Category.FreeGroupoid C
+  exact ⟨@default (Category.FreeGroupoid C) _⟩
+
+variable {C} in
+instance : Groupoid (Category.FreeGroupoid C) :=
+  Quotient.groupoid (Category.FreeGroupoid.homRel C)
+
+namespace FreeGroupoid
+
+@[simp]
+lemma of.map_id (X : C) : (Quotient.functor (FreeGroupoid.homRel C)).map
+    ((Groupoid.Free.of C).map (𝟙 X)) = 𝟙 _:= by
+  simp [Quotient.sound _ (Category.FreeGroupoid.homRel.map_id X)]
+
+@[simp]
+lemma of.map_comp {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    (Quotient.functor (FreeGroupoid.homRel C)).map ((Groupoid.Free.of C).map (f ≫ g)) =
+    (Quotient.functor (FreeGroupoid.homRel C)).map ((Groupoid.Free.of C).map f) ≫
+    (Quotient.functor (FreeGroupoid.homRel C)).map ((Groupoid.Free.of C).map g) := by
+  simp [Quotient.sound _ (Category.FreeGroupoid.homRel.map_comp f g)]
+
+/-- The localization map from the category `C` to the groupoid `Category.FreeGroupoid C` -/
+def of : C ⥤ Category.FreeGroupoid C where
+  __ := Groupoid.Free.of C ⋙q (Quotient.functor (FreeGroupoid.homRel C)).toPrefunctor
+  map_id X := by simp
+  map_comp {X Y Z} f g := by simp
+
+section UniversalProperty
+
+variable {C} {G : Type u₁} [Groupoid.{v₁} G]
+
+/-- The lift of a functor from `C` to a groupoid to a functor from
+`FreeGroupoid C` to the groupoid -/
+def lift (φ : C ⥤ G) : Category.FreeGroupoid C ⥤ G :=
+  Quotient.lift (FreeGroupoid.homRel C) (Groupoid.Free.lift φ.toPrefunctor) (by
+    intros X Y f g r
+    rcases r with X | ⟨ f , g ⟩
+    · simpa using Prefunctor.congr_hom (Groupoid.Free.lift_spec φ.toPrefunctor) (𝟙 X)
+    · have hf := Prefunctor.congr_hom (Groupoid.Free.lift_spec φ.toPrefunctor) f
+      have hg := Prefunctor.congr_hom (Groupoid.Free.lift_spec φ.toPrefunctor) g
+      have hfg := Prefunctor.congr_hom (Groupoid.Free.lift_spec φ.toPrefunctor) (f ≫ g)
+      simp only [Functor.toPrefunctor_obj, Prefunctor.comp_obj, Prefunctor.comp_map,
+        Functor.toPrefunctor_map, Quiver.homOfEq_rfl, Functor.map_comp] at *
+      rw [hf, hg, hfg])
+
+theorem lift_spec (φ : C ⥤ G) : of C ⋙ lift φ = φ := by
+  have : Groupoid.Free.of C ⋙q (Quotient.functor (FreeGroupoid.homRel C)).toPrefunctor ⋙q
+      (lift φ).toPrefunctor = φ.toPrefunctor := by
+    simp [lift, Quotient.lift_spec, Groupoid.Free.lift_spec]
+  fapply Functor.ext
+  · apply Prefunctor.congr_obj this
+  · intro _ _
+    simpa using Prefunctor.congr_hom this
+
+theorem lift_unique (φ : C ⥤ G) (Φ : Category.FreeGroupoid C ⥤ G) (hΦ : of C ⋙ Φ = φ) :
```

Comment:
Could you add the instance that `of C` is a localization functor (see `Functor.IsLocalization`) with respect to the class of all morphisms in `C`?